### DISCUSSION TITLE
IKVMC5045 Link Error

### DISCUSSION THREAD

**Discussion by mohitr-7** [ikvm-revived/ikvm-sdk-maven#64] [Q&A]
Hi - I'm not sure where exactly my problem lies, so posting this as a discussion rather than an issue.
I've created a Hello World console app (literally just used the default console project template VS provides), added the IKVM.Maven.Sdk v1.6.9 dependency via NuGet and then tried adding a maven dependency to the jt400 project.

If I add it as:
```
<ItemGroup>
  <MavenReference Include="jt400">
    <GroupId>net.sf.jt400</GroupId>
    <ArtifactId>jt400</ArtifactId>
    <Version>11.2</Version>
  </MavenReference>
</ItemGroup>
```
Then the project compiles fine (possibly worthy of note: jt400 v11.x doesn't have a pom, so not a proper maven project... I believe).

However, if I increment the version to the latest:
```
<ItemGroup>
  <MavenReference Include="jt400">
    <GroupId>net.sf.jt400</GroupId>
    <ArtifactId>jt400</ArtifactId>
    <Version>20.0.7</Version>
  </MavenReference>
</ItemGroup>
```
Then the build fails with this error:
```
fatal error IKVMC5045: Link error: Method "org.eclipse.sisu.plexus.PlexusBundlePlan$1.spaceModule()Lcom.google.inject.Module;" has a return type "com.google.inject.Module, sisu.guice, Version=3.2.0.0, Culture=neutral, PublicKeyToken=13235d27fcbfff58" and tries to override method "org.eclipse.sisu.launch.BundleModule.spaceModule()Lcom.google.inject.Module;" that has a return type "com.google.inject.Module, guice, Version=3.0.0.0, Culture=neutral, PublicKeyToken=13235d27fcbfff58"
```

This suggested something might be up with the updated jt400 package. But having downloaded the latest version of its code and doing a clean build, it doesn't have a reference to the org.eclipse.sisu project. I even cleared out my local maven folder to confirm it wasn't being downloaded from some deeply nested dependency. It only seems to get downloaded when trying to build the C# project.

So, am I missing something in my set up? Or is this some weird interaction that's breaking the IKVM compiler?

**Comment by wasabii** [comment]
Looks like it's trying to bring in the maven plugins from the POM file. Not sure why.

**Comment by wasabii** [comment]
So the reason is because https://repo1.maven.org/maven2/net/sf/jt400/jt400/20.0.7/jt400-20.0.7.pom lists org.apache.maven.plugins:maven-antrun-plugin as a dependency. I do not know why this is in here. But what it signals is that this jar is required to build the main jar. But not distributed at runtime. So, IKVM considers it to be a dependency required to transpile jt400, even though it should not be made available at runtime. Provided scope means the implementation classes will be present at runtime. But, at build time, the library mentioned should be used. So, IKVM follows through with that: it includes the plugin as it expects references to it in the classes at compile time.

I would bring this up with the jt400 people. Maven plugins should never be dependencies (unless you are actually writing a plugin).

**Reply by mohitr-7** [reply]
I tried changing the import to:
```
<ItemGroup>
  <MavenReference Include="jt400">
    <GroupId>net.sf.jt400</GroupId>
    <ArtifactId>jt400</ArtifactId>
    <Version>20.0.7</Version>
    <Exclusions>org.apache.maven.plugins:maven-antrun-plugin</Exclusions>
  </MavenReference>
</ItemGroup>
```
But I still get the same error. Have I got the syntax wrong, or would exclusions simply not help in this case?

(I'll try posting on the jt400 project too, but trying to see if there's a workaround in the meantime)

**Reply by mohitr-7** [reply]
Although, thinking about it further, the problem's really in the `maven-antrun-plugin` artifact rather than `jt400`, so probably need to report it higher up in the chain.

**Comment by mohitr-7** [comment]
After a bit of experimenting, this seems to get rid of the link error at least:
```
<ItemGroup>
  <MavenReference Include="jt400">
    <GroupId>net.sf.jt400</GroupId>
    <ArtifactId>jt400</ArtifactId>
    <Version>20.0.7</Version>
  </MavenReference>

  <MavenReference Include="maven-plugin-api">
    <GroupId>org.apache.maven</GroupId>
    <ArtifactId>maven-plugin-api</ArtifactId>
    <Version>3.9.8</Version>
  </MavenReference>
</ItemGroup>
```
`maven-antrun-plugin` references v3.2.5 of `maven-plugin-api`, and somewhere between v3.2.5 and v3.9.8 the `org.eclipse.sisu` discrepancy got addressed/replaced/removed.

There were a pile of class not found (IKVMC0100) and no such method (IKVMC0117) warnings, which may or may not be related. I guess I'll know once I try plugging it into something more than a Hello World.

**Comment by wasabii** [comment]
Probably. But including that plugin as a dep at all is a bug in jr400. You should report it

**Comment by mohitr-7** [comment]
Sure thing. Logged https://github.com/IBM/JTOpen/issues/190.

Do you think the `Exclusions` clause should've worked? I feel like it'd be a better workaround than the extra dependency on `maven-plugin-api`.

**Reply by wasabii** [reply]
It should work.

**Comment by wasabii** [comment]
Another thing here, for this version, is they have a specifically java8 artifact: https://repo1.maven.org/maven2/net/sf/jt400/jt400/20.0.7/

Notice the artifact ending in -java8.

You might try using that artifact instead. Which I believe would be changing the Classifier of the MavenReference to java8.

They seem to explicitly have a java8 and java11 artifact.... and then a default one. But I don't know what version of the JDK the default one is supposed to be targeting.

**Reply by mohitr-7** [reply]
Tried
```
<MavenReference Include="jt400">
  <GroupId>net.sf.jt400</GroupId>
  <ArtifactId>jt400</ArtifactId>
  <Classifier>java8</Classifier>
  <Version>20.0.7</Version>
</MavenReference>
```
(and `java11`) but no luck, it still triggers the link error without the additional `maven-plugin-api` reference :(
